### DISCUSSION TITLE
Typescript complains migrate function

### DISCUSSION THREAD

**Discussion by QQCif** [pmndrs/zustand#1142] [Q&A] [answered]
Hello,
I have created this codesandbox [https://codesandbox.io/s/awesome-wind-1rfmn9?file=/src/store/index.ts](https://codesandbox.io/s/awesome-wind-1rfmn9?file=/src/store/index.ts)
I have read #980, #1046 but I'm at a loss now. What's the proper way to write migrate function?

**Comment by dai-shi** [comment]
Here's something works: https://codesandbox.io/s/brave-bush-oe92xn?file=/src/store/index.ts

@devanshj Is this the only solution?

**Reply by QQCif** [reply]
Thank you.

**Comment by devanshj** [comment, marked as answer]
`persistedState` comes from the storage, hence it's type is `unknown` as it can be anything (it could be changed by the user, somewhere else in the code, XSS, etc), so this is intentional.

The correct way to solve this is do runtime checking, you can use something like [zod](https://github.com/colinhacks/zod) for that...

```ts
import create from "zustand";
import { devtools, persist } from "zustand/middleware";
import { z } from "zod";

interface DefaultStore {
  username?: string;
  age: number;
}

const useStore = create<DefaultStore>()(
  devtools(
    persist(
      (set) => ({
        username: undefined,
        age: -1
      }),
      {
        name: "default-state",
        version: 1,
        migrate: (persistedStateUnchecked, version) => {
          let persistedState: DefaultStore
          try {
            persistedState = z.object({
              username: z.string().optional(),
              age: z.number()
            }).parse(persistedStateUnchecked)
          } catch (e) {
            // handle what to do if the persisted state is not what we expect it to be
            // maybe you can tell user the problem, log them out and then clear the persisted storage
            // if it doesn't have anything important
            // maybe you can check what is missing from the persisted state and fill it up with some fallbacks
            // maybe just make it the initial state
            // what to do here depends on what is kept in the storage and why
            persistedState = { username: undefined, age: -1 }
          }
          
          if (version === 0) {
            persistedState.age = -1;
          }
          return persistedState;
        }
      }
    )
  )
);

export default useStore;
```

If you're confident that the storage will always have the current expected state or if you don't want to bother checking it, then in either case you can make a type assertion as @dai-shi showed...

```ts
migrate: (persistedStateUnchecked, version) => {
  let persistedState = persistedStateUnchecked as DefaultStore
  // ...
}
```

It's important that types are truthful hence zustand types `persistedState` as `unknown`, then it's upto users how they want to proceed from there.

Also maybe someone wants to include something like this in the [persist docs](https://github.com/pmndrs/zustand/blob/main/docs/persisting-store-data.md).

I also realised given this is `migrate` the runtime checking must also check for older schemas as `migrate` handles older schemas as well... So maybe if `age` is new in this new version of webapp/whatever then one should make it optional too and have some fallback for it.

**Reply by QQCif** [reply]
Thank you for this detail explanation. I have one more question, is it better to check all the fields from `persistedState` than to solely rely on version number, since it can also be tampered if the persist storage is `localStorage`.

**Reply by devanshj** [reply]
I personally would say that yes you can't rely on anything that comes outside the boundary of the program (localStorage, network, filesystem, etc). And you'd want to do runtime checks in `merge` too because `migrate` won't run if the version number is same.

Also to be clear I'm answering in my individual capacity :) I'm not author or maintainer of `persist`, so I'm not sure what kind of security concerns have been kept in mind while writing or what usage and practices zustand recommends.
